internal/driver/onvifdiscovery.go

```diff
@@ -0,0 +1,235 @@
+	return &OnvifProtocolDiscovery{driver: driver}
+}
+
+// ProbeFilter takes in a host and a slice of ports to be scanned. It should return a slice
+// of ports to actually scan, or a nil/empty slice if the host is to not be scanned at all.
+// Can be used to filter out known devices from being probed again if required.
+func (proto *OnvifProtocolDiscovery) ProbeFilter(_ string, ports []string) []string {
+	// For onvif we do not want to do any filtering
+	return ports
+}
+
+// OnConnectionDialed handles the protocol specific verification if there is actually
+// a valid device or devices at the other end of the connection.
+func (proto *OnvifProtocolDiscovery) OnConnectionDialed(host string, port string, conn net.Conn, params netscan.Params) ([]netscan.ProbeResult, error) {
+	// attempt a basic direct probe approach using the open connection
+	devices, err := executeRawProbe(conn, params)
+	if err != nil {
+		params.Logger.Debugf(err.Error())
+	} else if len(devices) > 0 {
+		return mapProbeResults(host, port, devices), nil
+	}
+	return nil, err
+}
+
+// ConvertProbeResult takes a raw ProbeResult and transforms it into a
+// processed DiscoveredDevice struct.
+func (proto *OnvifProtocolDiscovery) ConvertProbeResult(probeResult netscan.ProbeResult, params netscan.Params) (netscan.DiscoveredDevice, error) {
+	onvifDevice, ok := probeResult.Data.(onvif.Device)
+	if !ok {
+		return netscan.DiscoveredDevice{}, fmt.Errorf("unable to cast probe result into onvif.Device. type=%T", probeResult.Data)
+	}
+
+	discovered, err := proto.driver.createDiscoveredDevice(onvifDevice)
+	if err != nil {
+		return netscan.DiscoveredDevice{}, err
+	}
+
+	return netscan.DiscoveredDevice{
+		Name: discovered.Name,
+		Info: discovered,
+	}, nil
+}
+
+// createDiscoveredDevice will take an onvif.Device that was detected on the network and
+// attempt to get more information about the device and create an EdgeX compatible DiscoveredDevice.
+func (d *Driver) createDiscoveredDevice(onvifDevice onvif.Device) (sdkModel.DiscoveredDevice, error) {
+	xaddr := onvifDevice.GetDeviceParams().Xaddr
+	endpointRefAddr := onvifDevice.GetDeviceParams().EndpointRefAddress
+	if endpointRefAddr == "" {
+		d.lc.Warnf("The EndpointRefAddress is empty from the Onvif camera, unable to add the camera %s", xaddr)
+		return sdkModel.DiscoveredDevice{}, fmt.Errorf("empty EndpointRefAddress for XAddr %s", xaddr)
+	}
+	address, port := addressAndPort(xaddr)
+	dev := contract.Device{
+		// Using Xaddr as the temporary name
+		Name: xaddr,
+		Protocols: map[string]contract.ProtocolProperties{
+			OnvifProtocol: {
+				Address:            address,
+				Port:               port,
+				AuthMode:           d.config.DefaultAuthMode,
+				SecretPath:         d.config.DefaultSecretPath,
+				EndpointRefAddress: endpointRefAddr,
+			},
+		},
+	}
+
+	devInfo, edgexErr := d.getDeviceInformation(dev)
+	if edgexErr != nil {
+		// try again using the device name as the SecretPath
```

Comment:
Comment why need to do this

Reply:
This is kind of a hack that allows you to pre-fill the credentials per camera by using the endpoint ref. The upcoming story I am working on would replace this, so maybe I should just get rid of it for now? Or should I keep and just comment?

Reply:
Either works for me.